src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -102,13 +111,228 @@ private static async Task<VersionStamp> GetDocumentSemanticVersionAsync(Document
             return semanticVersion;
         }
 
-        // Internal for testing
-        internal async Task<SemanticTokens?> GetSemanticTokensAsync(
+        private async Task<SemanticTokens?> GetCachedTokensAsync(
             TextDocumentIdentifier textDocumentIdentifier,
             DocumentSnapshot documentSnapshot,
             int documentVersion,
             VersionStamp semanticVersion,
             Range range,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            if (!_cachedResponses.TryGetValue(textDocumentIdentifier.Uri, out var semanticVersionCache) ||
+                !semanticVersionCache.TryGetValue(semanticVersion, out var documentCache))
+            {
+                return null;
+            }
+
+            // 1) We'll first check if the cache contains an exact match for the range, in which case we can
+            // return early.
+            if (documentCache.TryGetValue(range, out var cachedTokens))
+            {
+                return cachedTokens;
+            }
+
+            // 2) If the cache doesn't contain an exact match for the range, we can still check if there's a
+            // partial match. This allows us to reduce the range we have to compute tokens for. We want to find
+            // the cached range with the most overlap with the requested range.
+            (Range Range, int NumOverlappingLines)? currentBestMatch = null;
+            foreach (var rangeCandidate in documentCache.Keys)
```

Comment:
I think this further solidifies that a `SemanticTokensCache` is the right way to go, and all the logic about caching should go in there.

Reply:
first, i agree.  this is very complex logic.  
second, i feel like trying to "Reduce teh range" is just not that helpful.  can we make it much simpler?   just check if the requested range falls within a range previously computed?

Reply:
note: if we do go this route, we should pull this into it's own type, and unit test heavily.  for example, i woudl expect tests that simply say "if this is what we cahced before, and we're asking for this new range, here's the actual subranges we need to classify".

Reply:
> Can you clarify this? what do you mean "finalized their tokens"?

C# can reply to Razor's request for Semantic Tokens with non-finalized tokens (Hasn't walked the whole C# tree yet). We need to not cache any response from C# which isn't finalized, otherwise we won't know when to request new tokens from C#. But we still want to return those results to users so that they get SOME tokens faster.